Added support for global meshes (stereographic projection)

tomsail opened the pull request:
Changes were made in the following routines: 

- [oceanmesh/__init__.py](https://github.com/CHLNDDEV/oceanmesh/pull/68/files#diff-96d5b2e3c46842ca76918d5e696e2d0a887de6f6721ed4cb2fa4ba167fffcce7): added the new functions in the path
- [oceanmesh/region.py](https://github.com/CHLNDDEV/oceanmesh/pull/68/files#diff-ae253938b5c07e054a98967bb740c58696f37d894df825990913f41d995181c3): added 4 function to pass from lat/lon to stereographic coordinates and vice versa
- [oceanmesh/edgefx.py](https://github.com/CHLNDDEV/oceanmesh/pull/68/files#diff-095e8eb6bb34032d1103ea536335504484a796ba4b6a805923d923f035910675): added supplementary gradient computation to eliminate warped and distorted values of the mesh size around the north (in stereographic projection)
- [oceanmesh/geodata.py](https://github.com/CHLNDDEV/oceanmesh/pull/68/files#diff-9790d791ac37a6f3c466c4b3cc0aebbd8e5642e04b027dd4e03e53ff50dac43d): rectified the `_AREAMIN` test (to suppress islands smaller than the resolution) for stereographic projection. 
- [oceanmesh/grid.py](https://github.com/CHLNDDEV/oceanmesh/pull/68/files#diff-b9634647edeeb3dbaa5a03542442f550bf8b2c38da7572fbf529c27123cdc84a): added stereo keyword in the arguments for the plot() function
- [oceanmesh/mesh_generator.py](https://github.com/CHLNDDEV/oceanmesh/pull/68/files#diff-56033f6d6a54d6ba402aae22c5ca4341c331f7a28ae5c360048f4212ffbad7d5): added the new function `_stereo_distortion` and the new keyword `stereo` to the options (`opts`) to rectify the following functions : 
  - _generate_initial_points(): initial distribution in the (-180.00, 180.00, -89.00, 90.00) bounding bbox and reproject them in stereo coordinates
  - _compute_forces(): rectify the computation of the lengths of the triangles segments, either due to the stereographic projection distortion or the elements accrooss the dateline (-180/+180)
- [tests/test_global_stereo.py](https://github.com/CHLNDDEV/oceanmesh/pull/68/files#diff-7cf86651d61a027f5caeda17836aa38af52dbadf51f6ec1c709f42c25abf8746): a test function that produces the following global mesh (0.5° resolution at the coast ~50km): 
![image](https://github.com/CHLNDDEV/oceanmesh/assets/18373442/d04b8dbf-82e7-4736-884d-a87e64f1f96c)

NB: bounding box for global mesh **has to** be (-180.00, 180.00, -89.00, 90.00)
NBB: it is necessary to define all the coastlines at once: the Shoreline class will the detect the biggest coastline (antartica and define it as the outside boundary). This is because the Shoreline hasn't been defined has a multi-shapefile reader - which would be a "nice to have"

tomsail commented:
I can also push the ocean.shp vector I used for the coastlines, as it is really light (42kB). Let me know

tomsail commented on oceanmesh/geodata.py:
```diff
@@ -210,13 +210,21 @@ def _classify_shoreline(bbox, boubox, polys, h0, minimum_area_mult):
+                bSGP = bSGP.difference(pSGP)
+                # Append polygon segment to mainland
+                mainland = np.vstack((mainland, poly))
+                # Clip polygon segment from boubox and regenerate path
```
true, I guess I could have just added the condition at the end and overwrite `bSGP`. 
Let me know what you think, I'll rectify it

tomsail commented:
For some reason (I'll need to justify it), after some sensibility tests (resolution, distance, feature and/or wavelength mesh sizes, this last version of the stereo distortion function gives the correct mesh on the north pole
![image](https://github.com/CHLNDDEV/oceanmesh/assets/18373442/c4f727ca-498d-47bb-97f5-fd9317f18122)

krober10nd commented:
@tomsail This is looking quite good. Could it be due to slight variations in the initial point distribution?

krober10nd commented on oceanmesh/edgefx.py:
```diff
@@ -122,6 +123,47 @@ def enforce_mesh_gradation(grid, gradation=0.15, crs="EPSG:4326"):
     cell_size = cell_size.flatten("F")
     tmp = gradient_limit([*sz], elen, gradation, 10000, cell_size)
     tmp = np.reshape(tmp, (sz[0], sz[1]), "F")
+    if stereo:
```
Is it possible to set the pole in the projection to somewhere over land to avoid this issue?

tomsail replied:
I will try it. In that case this implementation might be overkill

tomsail commented:
> @tomsail This is looking quite good. Could it be due to slight variations in the initial point distribution?

It could be. After the initial distribution - a grid covering the world with a resolution of `min_edge_length` - the excess points are randomly trimmed out using the mesh size. 
[I kept this functionality](https://github.com/CHLNDDEV/oceanmesh/pull/68/files#diff-56033f6d6a54d6ba402aae22c5ca4341c331f7a28ae5c360048f4212ffbad7d5L677), but also applied the stereographic distortion to adapt the selection in the transformed space.

tomsail commented:
@krober10nd I also have missed the stitching of the polygons on either side of the International Date Line (IDL): you can see clearly on the last picture I posted above the Russian peninsula cut from the rest of the continent. 

Summing up here is the checklist to complete this PR: 

- [ ] validate the stereographic distortion 
- [ ] evaluate the possibility to put the stereographic pole on land (probably Greenland) to avoid the extra gradient smoothing
- [ ] add minimal implementation for the stitching of world polygons (see pyposeidon function [global_tag](https://github.com/ec-jrc/pyPoseidon/blob/3dbcf5476dee1f3855fb7596a2c96c1d36558d52/pyposeidon/boundary.py#L448))

tomsail commented:
Hi @krober10nd, I corrected and simplified the compute_force() routine to account for the stereographic distortion. 

I am quite happy now with the quality of the mesh 
![Screenshot from 2023-11-08 17-58-45](https://github.com/CHLNDDEV/oceanmesh/assets/18373442/84f8d59e-781a-4c9f-b7c0-a34bf1e9c98c)

Except maybe on the poles: 
![image](https://github.com/CHLNDDEV/oceanmesh/assets/18373442/eed60d35-3742-4121-b2f9-4ca337024ef4)

This validate the first item of the checklist above. 

The other 2 items on the other hand would require both quite extensive implementations, which are already available in pyposeidon. 

2 options : 

1. We could merge it for now and address the other items in a further PR,
2. or wait for at least the 3rd item to merge this PR

krober10nd commented:
I will take a look at this hopefully this weekend thanks for your great work!!

krober10nd commented:
I've seen that convergence pattern in the mesh shape before and it has to do with the forces being too powerful. Could you try re-making the global meshes using a lower pseudo delta tilmestep smaller value? Also, the ratio of the medians for the scale factor could work better (in general) but that's not related to your changes. 

For example https://github.com/CHLNDDEV/OceanMesh2D/blob/f0c9257bc8f1a00836510e251328f07ef288324b/%40meshgen/meshgen.m#L811


I'm on board with merging this in, just please address the `flake8`  aspects and we're good to go.

krober10nd commented:
Seems like `scipy`'s interpolation routine now requires the coordinates to be montonically increasing causing most of the tests to fail.

tomsail commented:
Awesome Keith thanks. 
I'll fix the lint problems and try to propose also a workaround for the error (caused by `create_vectors()`)

tomsail commented:
I implemented 2 fixes and one enhancement, see the last commits: 

1. Monotonically ascending vectors. 
2. For the CIs, fixing the monotonically ascending vectors wasn't enough, there were outdated arguments in the `grid.plot()` function, I basically copied and adapted what was already there in the DEM.plot() function ) 

I also implemented the median rations (it was already implemented in the commented version of [`compute_forces()`](https://github.com/CHLNDDEV/oceanmesh/blob/1f9a52c1863527babd6a7dda928d663f329e4b5e/oceanmesh/mesh_generator.py#L599)  )

tomsail commented:
sorry I realised I didn't check the last implementations I did with flake8 ..

tomsail commented:
No I couldn't enhance or fix it

tomsail commented:
All CIs should pass now (I fixed flake8 and lowpass filter for bathy gradient)

krober10nd commented:
I will remove the bathymetric gradient test as this functional is not functional and needs to be revisited regardless.

tomsail commented:
FIY now the bathy gradient function works with the default `lowpass` filter. 
Reason for the function hanging was due to the ESPG problem (conversion degrees<>meters wasn't done at all) #69

tomsail commented:
@krober10nd, I will 

1. clean all the work in this PR (git rebase), 
2. check if the CIs work on my fork (I hadn't activated it before, now I do)
3. open it again for review once all CIs pass, so you can have a last look before merging

Note: 
I identified 2 issues that I will address in 2 different PRs for more clarity: 
1. The ESPG argument (mentioned above)
2. A bug in the bathy gradient through the grid.values setter

tomsail commented on oceanmesh/grid.py:
```diff
@@ -160,7 +160,7 @@ def create_vectors(self):
         x = self.x0y0[0] + np.arange(0, self.nx) * self.dx  # ascending monotonically
         y = self.x0y0[1] + np.arange(0, self.ny) * abs(self.dy)
-        y = y[::-1]  # descending monotonically
+        # y = y[::-1]  # descending monotonically
```
wait @krober10nd before merging. This induced a bug in the wavelength size function (now upside down). I need to rectify

krober10nd commented:
@tomsail everything looks good. could you add an example on the README with an image to highlight this new feature?

tomsail commented:
I added supporting files for global coastlines. It'll be clearer for users. 
Also this partly covers the following TODO: 

> * [x]  add minimal implementation for the stitching of world polygons (see pyposeidon function [global_tag](https://github.com/ec-jrc/pyPoseidon/blob/3dbcf5476dee1f3855fb7596a2c96c1d36558d52/pyposeidon/boundary.py#L448))

as I refer to the [`global_tag()`](https://github.com/ec-jrc/pyPoseidon/blob/9cfd3bbf5598c810004def83b1f43dc5149addd0/pyposeidon/boundary.py#L452) function in pyposeidon. Either in `test_global_stereo()` or in the `README`

I think this is a wrap
